runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/PipelineTranslation.java

```diff
@@ -102,11 +121,370 @@ public void visitPrimitiveTransform(Node node) {
       // TODO(JIRA-5649): Don't even emit these transforms in the generated protos.
       res = elideDeprecatedViews(res);
     }
+
+    ExternalTranslationOptions externalTranslationOptions =
+        pipeline.getOptions().as(ExternalTranslationOptions.class);
+    List<String> urnsToOverride = externalTranslationOptions.getTransformsToOverride();
+    if (urnsToOverride.size() > 0) {
+      // We use PTransformPayloadTranslators and the Transform Service to re-generate the pipeline
+      // proto components for the updated transform and update the pipeline proto.
+      Map<String, PTransform> transforms = res.getComponents().getTransformsMap();
+      List<String> alreadyCheckedURns = new ArrayList<>();
+      for (Entry<String, PTransform> entry : transforms.entrySet()) {
+        String urn = entry.getValue().getSpec().getUrn();
+        if (!alreadyCheckedURns.contains(urn) && urnsToOverride.contains(urn)) {
+          alreadyCheckedURns.add(urn);
+          // All transforms in the pipeline with the given urns have to be overridden.
+          List<
+                  AppliedPTransform<
+                      PInput,
+                      POutput,
+                      org.apache.beam.sdk.transforms.PTransform<? super PInput, POutput>>>
+              appliedPTransforms =
+                  findAppliedPTransforms(
+                      urn, pipeline, KnownTransformPayloadTranslator.KNOWN_PAYLOAD_TRANSLATORS);
+          for (AppliedPTransform<
+                  PInput,
+                  POutput,
+                  org.apache.beam.sdk.transforms.PTransform<? super PInput, POutput>>
+              appliedPTransform : appliedPTransforms) {
+            TransformPayloadTranslator<
```

Comment:
Rather than re-look this up, I still think it'd be cleaner to attach the Row to the PTransform itself during translation (at the same time you're attaching the URN and doing the translation, so you already have a PayloadTranslator in hand), and then just walk over the graph looking for URNs that you need to swap out. 

You can stick the config row (and its schema) in https://github.com/apache/beam/blob/master/model/pipeline/src/main/proto/org/apache/beam/model/pipeline/v1/beam_runner_api.proto#L207